src/Features/LanguageServer/Protocol/ExternalAccess/Razor/GetProjectContextsHandler.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+using System.IO;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.ExternalAccess.Razor;
+
+[ExportCSharpVisualBasicStatelessLspService(typeof(GetProjectContextsHandler)), Shared]
+[Method(GetProjectContextsName)]
+internal class GetProjectContextsHandler : ILspServiceDocumentRequestHandler<ProjectContextsParams, ProjectContextList?>
+{
+    private const string GetProjectContextsName = "roslyn/getProjectContexts";
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public GetProjectContextsHandler()
+    {
+    }
+
+    public bool MutatesSolutionState => false;
+
+    public bool RequiresLSPSolution => true;
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(ProjectContextsParams request)
+        => request.TextDocument;
+
+    public Task<ProjectContextList?> HandleRequestAsync(ProjectContextsParams request, RequestContext context, CancellationToken cancellationToken)
+    {
+        Contract.ThrowIfNull(context.Workspace);
+        Contract.ThrowIfNull(context.Solution);
+
+        var contextList = ProjectContextHelper.GetContextList(context.Workspace, context.Solution, request.TextDocument.Uri);
+        if (contextList is null)
+        {
+            return SpecializedTasks.Null<ProjectContextList>();
+        }
+
+        var idToIntermediateOutputMap = new Dictionary<string, string?>();
+        foreach (var projectContext in contextList.ProjectContexts)
+        {
+            if (projectContext is null)
+            {
+                continue;
+            }
+
+            var projectId = ProtocolConversions.ProjectContextToProjectId(projectContext);
+            var project = context.Solution.GetRequiredProject(projectId);
+
+            var dllOutputPath = project.CompilationOutputInfo.AssemblyPath;
+            Contract.ThrowIfNull(dllOutputPath);
+
+            var outputDirectory = Directory.GetParent(dllOutputPath)!.FullName;
```

Comment:
Nit: Can just use `Path.GetDirectoryName`, which is what `Directory.GetParent` seems to use, and avoid constructing a `DirectoryInfo` object

Reply:
This also seems very odd to me that we'd strip out the path like this when I imagine you could also do that same filtering somewhere else? I guess I'm surprised we key on _output directory_ rather than just path.

Reply:
Or put another way: if you can do this same operation on the other side, should you so this is more generally useful?

Reply:
I always thought `Path.GetDirectoryName` didn't return the full path. Looks like I was wrong :)